### ISSUE TITLE
podAntiAffinity

### ISSUE THREAD

**Issue by joebowbeer** [aws/karpenter#942]
**Tell us about your request**

Please implement `podAntiAffinity`.

I don't see anything about this on the roadmap. Is this planned?

(In fact, I see almost nothing on the [roadmap](https://github.com/aws/karpenter/projects/3) currently. Issue?)

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**

I want to deploy [k8ssandra](https://k8ssandra.io/) but cannot because it is using pod anti-affinity to prevent multiple pods from being scheduled on the same worker

> Cass Operator uses pod anti-affinity by default to ensure that Cassandra pods are isolated from one another.

and Karpenter refuses to schedule.

I suspect there are a lot of high-availability operators that will currently _prefer_ or _require_ podAntiAffinity.

Prometheus: https://github.com/prometheus-operator/kube-prometheus/issues/1537

**Are you currently working around this issue?**

`allowMultipleNodesPerWorker: true`

From: https://docs.k8ssandra.io/tasks/scale/

**Additional context**

What should these operators be using instead?

I see in the FAQ that this is not a priority for Karpenter but I think podAntiAffinity is fairly common among some popular operators, and it is not clear what they should be doing instead.

Some discussion here: https://github.com/aws/karpenter/pull/695#discussion_r715817587

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by akestner** [comment]
Apologies for the lack of detail on the roadmap, we're working to get that updated to reflect the next steps for Karpenter after the launch at re:Invent last week. 

Regarding `podAntiAffinity`, this is something we've discussed in the past, but it presents a number of challenges to Karpenter's scheduling capabilities. Often, as seems to be the case with k8ssandra and the other projects mentioned, topology spread could be a suitable alternative. I'll let @ellistarn weigh in as well since I know he's thought quite a bit about this too.

**Comment by joebowbeer** [comment]
@akestner Thanks for the background.

I suspect several CNCF projects are configured with _soft_ podAntiAffinity now.

For example, these [several in prometheus-operator](https://github.com/prometheus-operator/kube-prometheus/blob/main/jsonnet/kube-prometheus/addons/anti-affinity.libsonnet#L2-L19), and it's not clear to everyone that this configuration is inappropriate. (See [discussion](https://github.com/prometheus-operator/kube-prometheus/issues/1537) linked in description.)

In the interim, can Karpenter implement soft podAntiAffinity as if it were topology spread?

**Comment by ellistarn** [comment]
>In the interim, can Karpenter implement soft podAntiAffinity as if it were topology spread?

This is definitely how we'd do it. We already mutate pods in memory for similar things (i.e. nodeselectors are modeled as nodeaffinity in memory). The spec is very broad -- I can see support for a limited set of pod anti-affinity and make sure that maps well to common charts like kassandra. 

Pod affinity (not anti) is extremely challenging. I don't know how to do it in a reasonable way -- it doesn't really make sense for provisioned clusters. I'd need to see an airtight technical proposal before we could make progress here.

**Comment by bdwyertech** [comment]
EKS built-in CoreDNS leverages a pod anti-affinity rule.

Stumbled across this wondering why Karpenter is spinning up nodes in the same region with a topology zone anti-affinity rule configured 😄  Many Helm charts expose affinity but do not expose topologySpreadConstraints.

+1

**Comment by Noksa** [comment]
I've added some explanation why we need `PodAntiAffinity`:
https://github.com/aws/karpenter/issues/1010#issuecomment-1011171630

**Comment by SamPeng87** [comment]
This is a very popular feature to use, and if karpenter is deployed you have to pay for it, such as calculating a new machine to put in first from request. This is the simplest scheduling strategy, but it solves 90% of the problems.

**Comment by jlarfors** [comment]
Seems like almost every 3rd party Helm chart has a default affinity of `podAntiAffinity`, and doesn't expose `topologySpreadConstraints`... This really should be a high priority if Karpenter is to be a real alternative to Cluster AutoScaler, which I think it should be as Karpenter is awesome otherwise.

**Comment by ellistarn** [comment]
Doing a bit of research -- it looks like podAffinity is not supported for daemonsets: https://github.com/kubernetes/kubernetes/issues/29276. This should make our implementation a little easier.

**Comment by ellistarn** [comment]
Further, looking at https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity:

> For requiredDuringSchedulingIgnoredDuringExecution pod anti-affinity, the admission controller LimitPodHardAntiAffinityTopology was introduced to limit topologyKey to kubernetes.io/hostname. If you want to make it available for custom topologies, you may modify the admission controller, or disable it.


Initial support may be constrained to `kubernetes.io/hostname`.

**Comment by ellistarn** [comment]
It looks like we may be able to reuse our topology implementation for anti-affinity. I have a prototype here: https://github.com/aws/karpenter/pull/1221/files

**Comment by joebowbeer** [comment]
@ellistarn can you clarify what these namespace restrictions mean in practice?

For example, would they still support the antiAffinity uses in the Prometheus Operator? (https://github.com/aws/karpenter/issues/942#issuecomment-989075615)

I'm not sure I follow your statement that best practice is to avoid use of these features. By features do you mean pod affinity/antiAffinity? The only statement I find is that it is best to avoid them in clusters larger than several hundred nodes.

**Comment by ellistarn** [comment]
Cross namespace affinity gets expensive quickly. IIUC the prometheus operator is using `namespaces: [prometheus]`. This is how topology works by default (i.e. same namespace), so its trivial to open it up to that. Supporting cross namespace is definitely a bigger lift.

**Comment by joebowbeer** [comment]
> IIUC the prometheus operator is using `namespaces: [prometheus]`. This is how topology works by default (i.e. same namespace), so its trivial to open it up to that.

Yes, that is my reading of the code as well. Thanks for clarifying.

**Comment by ellistarn** [comment]
See https://github.com/aws/karpenter/pull/1221/files#diff-dd2ae84838dbbe5f8348c698c96aecf8d9ee688821c432568a42029259e29474R48

**Comment by cheesimp** [comment]
The PR opened is for "hostname" but we are using "zone". Can we get this added there too?

**Comment by zswanson** [comment]
In the meantime could the log about this get bumped from DEBUG to ERROR? I have been pulling my hair out wondering why I had pods stuck unscheduled for hours until I changed the log level on karpenter.

**Comment by ellistarn** [comment]
Good suggestion. https://github.com/aws/karpenter/pull/1423/files

**Comment by matthiasdotbertschattngtechdotcom** [comment]
We also stumbled across this in combination with Tekton's [Affinity Assistant](https://tekton.dev/vault/pipelines-v0.16.3/workspaces/#specifying-workspace-order-in-a-pipeline-and-affinity-assistants). It uses inter-pod affinity and anti-affinity to schedule `TaskRun`s that share a `Workspace` to the same node by using an affinity assistant pod (with an anti-affinity to other affinity assistant pods). Currently, Karpenter simply refuses to schedule these pods.

Support for `podAntiAffinity` would also help here.

**Comment by cebernardi** [comment]
We also are encountering this issue, but with `podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution`.
I was wondering what the community thoughts are about relaxing these soft podAntiAffinity configurations, as it was done for `nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution`: https://github.com/aws/karpenter/blob/ba89ba83cb8f67c726dea0a1a0ba2ea57bbe1787/pkg/controllers/selection/preferences.go#L86-L99

I'm not familiar enough with the project pillars, but from a user point of view it feels like _preferred_ affinities shouldn't
t prevent pods from being scheduled. 

What are the maintainer's thoughts? Can we help with this?

**Comment by ellistarn** [comment]
I would absolutely accept a PR that does this.

**Comment by adjain131995** [comment]
I can see this issue being raised for quite a while and it is a very demanding feature. Do we have any ETA when this will be implemented in karpenter. I do not see any other alternative which we could use and topology constraints are hard ones and the alternatives tried also fails with kube0schedduler kicking in. Here is the track of that one

#1528

**Comment by adjain131995** [comment]
Also in my use case i would like to use pod anti-affinity 
`      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          -  podAffinityTerm:
               labelSelector:
                 matchLabels:
                   app.kubernetes.io/name: {{ .Release.Name }}
               topologyKey: kubernetes.io/hostname
             weight: 100`
to ensure pods prefer to spread and not go on the same node where that are running but in case there is no option they get deployed. Only way to counter this in karpenter is using topologyKey: kubernetes.io/hostname which forces karpenter to spins up a huge number of nodes in case there is such a requirement.
How can this be catered in karpenter with using pod affinity/anti-affinity

**Comment by cebernardi** [comment]
@ellistarn I'm taking a stab

**Comment by ellistarn** [comment]
Should be straightforward here: https://github.com/aws/karpenter/blob/main/pkg/controllers/selection/preferences.go#L51

**Comment by cebernardi** [comment]
@ellistarn https://github.com/aws/karpenter/pull/1541

**Comment by cebernardi** [comment]
ops, I saw your comment now, I took another approach, please let me know your thoughts

**Comment by ellistarn** [comment]
Either works. Removing in preferences will survive when we implement affinity for real.

**Comment by cebernardi** [comment]
I can include this in the PR of course!
Would you suggest removing them iteratively as it's done for podAffinity, or removing every terms at once?

**Comment by ellistarn** [comment]
Iteratively SGTM

**Comment by marcportabellaclotet-mt** [comment]
It would be great is this feature (antiaffinity) is supported by karpenter.
Our use case is jenkins-operator. We spread every job to a new node using antiaffinity, and it there is no jenkins job, nodes scale to 0.
Karpenter is an Amazing project!!

**Comment by tzneal** [comment]
If you've got a fairly recent version of Karpenter already running, you can drop in this snapshot controller image to test out pod affinity & pod anti-affinity.  This is just a snapshot, not meant for production, etc.:

```
kubectl set image deployment/karpenter -n karpenter controller=public.ecr.aws/z4v8y7u8/controller:08ac9ec303aabe2da56edb0ee0e235a60a287206
```

**Comment by iaacautomation** [comment]
Its also useful for elasticserach operator, if you create 3 node cluster topology spread can schedule 2,1 while preferred for HA is 1,1,1.

**Comment by tzneal** [comment]
@iaacautomation Pod Affinity & Anti-Affinity support was released in v0.9.0.  Feel free to open a different issue if you are having problems with them.
